xdsl/transforms/memref_stream_legalize.py

```diff
@@ -0,0 +1,150 @@
+from dataclasses import dataclass
+from typing import TypeAlias, cast
+
+from xdsl.context import MLContext
+from xdsl.dialects import memref_stream
+from xdsl.dialects.builtin import (
+    ArrayAttr,
+    Float16Type,
+    Float32Type,
+    Float64Type,
+    IntegerAttr,
+    ModuleOp,
+    VectorType,
+    f16,
+    f32,
+)
+from xdsl.dialects.linalg import IteratorType
+from xdsl.ir import Attribute
+from xdsl.passes import ModulePass
+from xdsl.pattern_rewriter import (
+    GreedyRewritePatternApplier,
+    PatternRewriter,
+    PatternRewriteWalker,
+    RewritePattern,
+    op_type_rewrite_pattern,
+)
+from xdsl.utils.exceptions import DiagnosticException
+from xdsl.utils.hints import isa
+
+StreamingVectorType: TypeAlias = VectorType[Float64Type | Float32Type | Float16Type]
+StreamingLegalType: TypeAlias = StreamingVectorType | Float64Type
+
+
+def _is_legal_vector(attr: Attribute) -> bool:
+    if isa(attr, StreamingVectorType):
+        match attr.element_type, attr.element_count():
+            case Float64Type(), 1:
+                return True
+            case Float32Type(), 2:
+                return True
+            case Float16Type(), 4:
+                return True
+            case _:
+                return False
+    return False
+
+
+def _is_legal_attr(attr: Attribute) -> bool:
+    return _is_legal_vector(attr) or isa(attr, StreamingLegalType)
+
+
+def _legalize_attr(attr: Attribute) -> StreamingLegalType:
+    if isa(attr, StreamingVectorType):
+        # Either already a legal vector or impossible to legalize
+        if not _is_legal_vector(attr):
+            raise DiagnosticException(f"Cannot legalize {attr} for streaming")
+        return attr
+    elif isinstance(attr, Float64Type):
+        return attr
+    elif isinstance(attr, Float32Type):
+        return VectorType(f32, (2,))
+    elif isinstance(attr, Float16Type):
+        return VectorType(f16, (4,))
+    else:
+        raise DiagnosticException(f"Cannot legalize {attr} for streaming")
+
+
+@dataclass(frozen=True)
+class MemrefStreamGenericLegalize(RewritePattern):
+
+    @op_type_rewrite_pattern
+    def match_and_rewrite(
+        self, op: memref_stream.GenericOp, rewriter: PatternRewriter
+    ) -> None:
+        # FIXME add reduction support
+        if op.iterator_types.data[-1].data != IteratorType.PARALLEL:
+            return
```

Comment:
I would say we should actually raise a DiagnosticError and crash early right now, since we're planning to add it soon